broadinstitute/single_cell_portal_core
# Filter IGV by cell annotations, tailor ATAC rendering (SCP-5612, SCP-5614)

This enables filtering IGV features by cell annotations, and refines how IGV shows ATAC.  It enriches multiome exploration.

Here's how it looks!

https://github.com/broadinstitute/single_cell_portal_core/assets/1334561/f8060e12-75d6-4a8a-89ea-f5b41b3bca2f

(Filtering is slow.  It's behind a flag; next steps include optimization.)

### Impact
This lets users see how chromatin accessibility differs in _custom-filtered_ cell populations, in genomic context.  So if your multiome data has e.g. cell type annotations, you can see how transcribable any gene of interest is in cells filtered to show only e.g. B cells or macrophages.  

Because it's multiome, these genomic insights can be related to expression insights for the same _filtered_ cell population, e.g. plots showing UMAP clusters and categorical distributions.

### Overview
[Previously](https://github.com/broadinstitute/single_cell_portal_core/pull/2041), users could upload scATAC-seq fragments as BED files and view them in igv.js.  However, our IGV rendered ATAC data as generic BED, which obscured relevant scientific details.  It also didn't integrate with powerful SCP exploration features.  ATAC fragment BEDs in IGV integrated with gene search, but not cell filtering.

Now, multiome IGV has better default rendering, and integrates cell filtering from an earlier [proof of concept](https://github.com/broadinstitute/single_cell_portal_core/pull/2021).  This also removes the "Genome" tab from the unfiltered view, but retains it in the single-gene view.

#### Tailored rendering
Per [discussion](https://broadinstitute.slack.com/archives/CESEYJW9W/p1717099893194269) with the Pipeline Development team, instead of coloring ATAC fragments by BED "score" with a random palette, fragment colors are now meaningful.  A sequential palette colors fragments with higher "scores" increasingly bold red.  The popover shown on clicking ATAC features now labels field names more specifically: "Name" is "ATAC barcode", and "Score" is "Read count".  Other IGV feature properties are refined for easier exploration.

#### Cell filtering in IGV
Cell filtering for genomic features in IGV is also now available in the Explore UI, when the new `show_igv_multiome` feature flag is on.  It works at a rudimentary level.  The filtering uses cell name indexes that extend those for violin plots.  For multiome data, cell barcodes for ATAC fragments match the cell barcodes for GEX data, so applying filters now not only updates cells shown for GEX data in scatter plots and violin plots, but also cells shown for ATAC data in IGV.

SCP's fork of igv.js was updated to account for this.  Code for that is in https://github.com/broadinstitute/igv.js/tree/dynamic-filtering.  Once IGV filtering is optimized, the plan is to open a PR upstream to merge it into the canonical igv.js.

### Next steps
* Speed it up.  Taking 10-30 seconds to filter is drastically faster than writing code, running it, and uploading new tracks.  But such speeds are painfully slow in a web app.  Filtering off the main thread, and faster, are part of SCP-5662.

* Make it more robust.  IGV filtering is brittle at this prototype stage.  For example, if you search for a new gene, things go awry.  Handling this common case is also part of SCP-5662.  That ticket seems reasonable for any other non-preexisting high-impact SCP Explore IGV issues that only manifest with the `show_igv_multiome` feature flag enabled.

* Show peaks from piles (SCP-5635).  This will make ATAC fragment distributions glanceable _and filterable_, while still enabling inspection of particular fragments.

Additional future work could include enabling custom color scales and track names, better handling for non-ATAC BED files, and handling multiple BED files.

### Tests
A new automated test confirms IGV filtering.  To manually test:

1.  Check out this feature branch and run `rails db:migrate` (only needs to be done once)
2. Set the `show_igv_multiome` feature flag to `true` 
3. Copy [SCP395 on staging](https://singlecell-staging.broadinstitute.org/single_cell/study/SCP395) to your local SCP
4. Download ATAC fragment BED bundle (i.e., both files) from [this bucket location](https://console.cloud.google.com/storage/browser/fc-bcc55e6c-bec3-4b2e-9fb2-5e1526ddfcd2/reference_data_dev/nemo_atac/10x_multiome_human_pbmc)
5. Upload that BED bundle to your local SCP
3. In your study with an ATAC fragment file, search a gene
6. Confirm "Genome" tab appears
7. Click Genome tab, confirm IGV appears
8. Confirm track with features for BED file appears, with features colored grey to bold red
9. Click on any BED feature (i.e., narrow rectangle)
10. Confirm popover shows values for "Read count" and "ATAC barcode"
11. Click "Filter plotted cells"
12. Deselect a filter
13. Wait a while; you might even need to click "Wait" if Chrome shows an unresponsive script prompt.  (Optimizing performance is a next step.)
14. Confirm BED track (eventually) updates with fewer features

### Further details
This satisfies SCP-5612 and SCP-5614.  More context is available in the [2024-06-04 SCP demo video](https://drive.google.com/file/d/1vroZOW8gEa7NXiRjhCO-3q8FbqVh-A52/view?t=29).

This also re-deletes `ui_test_suite.rb`, which was accidentally restored while updating [PR 2042](https://github.com/broadinstitute/single_cell_portal_core/pull/2042).
